File: Kawai codes/16.Random-Walk/ch16pr02.m
%**************************************************************************
%*     Exercise 16.2                                                      *
%*     filename: ch16pr02.m                                               *
%*     program listing number: 16.2                                       *
%*                                                                        *
%*     This program simulates the one-dimensional persistent random walk. *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course.       *
%*     Last modification:  03/04/2017.                                    *
%**************************************************************************
clear all;

% parameters
p=0.75;  % persistent jump probability
M=100000; % number of particles
N=1000; % number of steps

% initial states
x=zeros(M,1);
mu=zeros(N,1);
sigma2=zeros(N,1);
mu(1)=0.;
sigma2(1)=0.;

d=randi(2,[M,1])*2-3; %unbiased initial jump
x=x+d;
mu(2)=sum(x)/M;  % mean
sigma2(2)=sum(x.^2)/M-mu(2)^2; %variance

for i=3:N
    r=rand(M,1);
    k=r>p;  % direction is reversed.
    d(k)=-d(k);
    x=x+sign(d);  % jump
    mu(i)=sum(x)/M;  % mean
    sigma2(i)=sum(x.^2)/M-mu(i)^2; %variance
end

subplot(1,2,1)
p=plot([1:N],mu,[1:N],sigma2);
set(p(1),'linewidth',2,'color','blue')
set(p(2),'linewidth',2,'color','red')
hold on
p=plot([1:N],[1:N],'--');
set(p,'color','black')
legend('\mu','\sigma^2','\sigma^2 (Normal RW)')
legend('location','northwest')
xlabel('steps')
ylabel('moments')
hold off

subplot(1,2,2)
h=histogram(x,51,'Normalization','pdf');
y=h.BinEdges;
g=1.0/sqrt(2*pi*N)*exp(-y.^2/(2.0*N));
hold on
r=plot(y,g);
set(r,'color','red')
xlabel('x')
ylabel('probability density')
legend('persistent RW','normal RW')
hold off

File: Kawai codes/16.Random-Walk/ch16pr02.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
%**************************************************************************
%*     Exercise 16.2                                                      *
%*     filename: ch16pr02.m                                               *
%*     program listing number: 16.2                                       *
%*                                                                        *
%*     This program simulates the one-dimensional persistent random walk. *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course.       *
%*     Last modification:  03/04/2017.                                    *
%**************************************************************************
"""
import numpy as np
import matplotlib.pyplot as plt

# parameters
p=0.75   # persistent jump probability
M=100000  # number of particles
N=1000    # number of steps

mu=np.zeros(N+1)
sigma2=np.zeros(N+1)
t=np.linspace(0,N,N+1)

# initial states
x=np.zeros(M)  # reset the trajectories
mu[0]=0.
sigma2[0]=0.

d=np.random.choice([-1,1],M)
x=x+d # unbiased initial jump
mu[1]=np.sum(x)/M        # mean
sigma2[1]=np.sum(x**2)/M-mu[1]**2  #variance

for i in range(1,N+1):
    r=np.random.rand(M)
    k=r>p  # direction is reversed.
    d[k]=-d[k]
    x=x+np.sign(d)        # jump
    mu[i]=np.sum(x)/M;       # mean
    sigma2[i]=np.sum(x**2)/M  # ariance

plt.figure(figsize=(12,5))
plt.subplot(1,2,1)
plt.plot(t,mu,'-b',label=r'$\mu$',linewidth=2)
plt.plot(t,sigma2,'-r',label=r'$\sigma^2$',linewidth=2)
plt.plot(t,t,'--k',label='$\sigma^2$ (Normal RW)')
plt.legend(loc=2)
plt.xlabel('steps')
plt.ylabel('moments')

plt.subplot(1,2,2)
n, X, Y = plt.hist(x,51,normed=True,label='persistent RW')
dX=X[2]-X[1]
Z=1.0/np.sqrt(2.0*np.pi*N)*np.exp(-X**2/(2.0*N))
plt.plot(X,Z,'-r',label='normal RW')
plt.xlabel('x')
plt.ylabel('probability distribution')
plt.legend(loc=1)
plt.show()
